atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java

```diff
@@ -48,11 +50,25 @@
                 .orElseGet(ImmutableList::of);
     }
 
-    LockWatchEvents retentionEvents() {
-        int numToRetention = Math.max(0, eventMap.size() - maxEvents);
+    LockWatchEvents retentionEvents(Optional<Long> earliestSequenceToKeep) {
+        if (eventMap.size() < maxEvents) {
+            return LockWatchEvents.builder().build();
+        }
+
+        int numToRetention = eventMap.size() - maxEvents;
+        long earliestVersion = earliestSequenceToKeep.orElse(Long.MAX_VALUE);
+
         ImmutableLockWatchEvents.Builder builder = LockWatchEvents.builder();
-        Iterators.consumingIterator(Iterators.limit(eventMap.entrySet().iterator(), numToRetention))
-                .forEachRemaining(entry -> builder.addEvents(entry.getValue()));
+        List<Map.Entry<Long, LockWatchEvent>> eventsToClear = eventMap.entrySet().stream()
+                .limit(numToRetention)
```

Comment:
I know it technically does not matter since the treemap will return entries in ascending key order, but still let's swap the filter and limit so that it would also work with a generic map.

Reply:
I disagree; if no version is provided, then swapping the filter and the limit does nothing, and it still depends on the behaviour of the sorted map (which it did before in this class). It would actually just be wrong to have a generic map here (the type is still `NavigableMap` which forces this).

Reply:
I mean I see what you're trying to say, even though you are not saying it explicitly, which is that we want to guarantee that we always retention from the oldest, so if our `Set` here was not sorted, we'd be in trouble. Also, I guess this ordering of limit and filter does limit the amount of filtering we do, although we could further improve this by having a loop instead of a stream. However, this requires us to jump through mental hoops because it's too smart: we have an `entrySet()` of some map, then we take `numToRetention` elements and do the logic. Wait, it's a set -- better go make sure that `eventMap.entrySet()` returns a set that is sorted in the correct way for this logic to work. Oh, it's a `NavigableMap`, well that still doesn't guarantee it. Oh, it's actually initialized to a `TreeMap()`, ok let's go see the javadoc. Ok, I guess it works.

In fact, the performant way to do this would be to just have  a for loop and break when either condition is met, but that would look ugly, so I can live with adding a comment above this stream reminding us that this entryset is sorted ascending by key

Reply:
I actually think the for loop would be a bit clearer in what the semantics are, so I think I'll go with that, and also add javadocs.

Reply:
I'll comment here (although `NavigableMap` does say on its javadoc via `SortedMap` that it _does_ guarantee the keys are sorted in `entrySet`).